src/sensors/sensor.ts

```diff
@@ -0,0 +1,171 @@
+}
+
+export interface IPollOptions {
+  firstPoll?: boolean;
+  lastPoll?: boolean;
+}
+
+export class Sensor extends EventEmitter<SensorEvent> {
+  protected _deviceName: string | undefined;
+  private _connected: boolean;
+  private _values: ISensorValues;
+  private _capabilities: ISensorCapabilities;
+  private pollTimeout: number;
+  private pollInterval: number;
+  private error: any;
+
+  constructor(options: ISensorOptions) {
+    super();
+    this._capabilities = options.capabilities;
+    this._connected = false;
+    this._values = {};
+    this.pollInterval = options.pollInterval || 1000;
+    this.error = undefined;
+  }
+
+  public get capabilities() {
+    return this._capabilities;
+  }
+
+  public get connected() {
+    return this._connected;
+  }
+
+  public get values() {
+    return this._values;
+  }
+
+  public get deviceName() {
+    return this._deviceName || (this._connected ? "Unknown Device" : undefined);
+  }
+
+  public connect(): Promise<void> {
+    return Promise.reject("connect() method not overridden!");
+  }
+
+  public disconnect(): Promise<void> {
+    return Promise.reject("disconnect() method not overridden!");
+  }
+
+  public setError(error: any) {
+    // only set the first error to help in debugging or allow unset
+    if (!this.error || !error) {
+      this.error = error;
+      const errorData: ISensorErrorData = {
+        deviceName: this.deviceName,
+        error
+      };
+      this.emit(SensorEvent.Error, errorData);
+    }
+  }
+
+  protected pollValues(options: IPollOptions): Promise<ISensorValues> {
+    return Promise.reject("pollValues() method not overridden!");
+  }
+
+  // can't use private setter as it must agree with public getter
+  protected setConnected(options: ISetConnectedOptions) {
+    const {connected, deviceName} = options;
+    const wasConnected = this._connected;
+    this._connected = connected;
+    const connectionEventData: ISensorConnectionEventData = {
+      connected,
+      deviceName: deviceName || this.deviceName
+    };
+
+    clearTimeout(this.pollTimeout);
+    if (connected) {
+      this._deviceName = deviceName;
+      this.poll({firstPoll: true});
+    } else {
+      if (wasConnected) {
+        this.poll({lastPoll: true});
+      }
+      this._deviceName = undefined;
+    }
+
+    this.emit(SensorEvent.Connection, connectionEventData);
+  }
+
+  private poll(options: IPollOptions) {
+    const createTimeout = () => {
+      this.pollTimeout = window.setTimeout(() => {
+        this.sendData(options).then(() => this.poll({firstPoll: false}));
+      }, this.pollInterval);
+    };
+
+    if (options.firstPoll) {
+      // immediately send the data and then queue another poll
+      this.sendData(options).then(createTimeout);
+    } else if (options.lastPoll) {
+      // immediately send the data and then stop
+      this.sendData(options);
+    } else {
+      // queue another poll if not first or last poll
+      createTimeout();
+    }
+  }
```

Comment:
I setup the sensor code so that devices only need to expose a limited set of functions that get called from the sensor "manager".  Device polling uses the following sequence of public sensor methods `setupRead` -> `read` -> `teardownRead` which are controlled by those two booleans - maybe a better way to do it would be to have a state variable with values `setup` | `read` | `teardown`.  I will probably fiddle with the sensor code to get the other device added this week so this code is still a little in flux but I thought it would be good to get it merged so you could use it for the sensor table (the component shouldn't change).